src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,500 @@
+   * @param row to be serialized
+   * @param schema schema of the row
+   * @param proj_map projection map of the schema.
+   */
+  void SerializeTuple(ProjectedRow *row, const TupleSlot *slot, const catalog::Schema &schema,
+                      const ProjectionMap &proj_map);
+
+  /**
+   * Get the current checkpoint page.
+   * @return pointer to the checkpoint page header.
+   */
+  CheckpointFilePage *GetPage() { return reinterpret_cast<CheckpointFilePage *>(buffer_); }
+
+  /**
+   * Set the cached table oid. This value will be written to every page, unless otherwise changed.
+   * @param oid of the table
+   */
+  void SetTableOid(catalog::table_oid_t oid) {
+    table_oid_ = !oid;
+    GetPage()->SetTableOid(oid);
+  }
+
+  /**
+   * Set the cached table schema version. This value will be written to every page, unless otherwise changed.
+   * Currently the table only have one version, so this should be 0 all the time.
+   * @param version of the table schema
+   */
+  void SetVersion(uint32_t version) {
+    version_ = version;
+    GetPage()->SetVersion(version);
+  }
+
+ private:
+  AsyncBlockWriter async_writer_;  // background writer
+  uint32_t block_size_;
+  uint32_t page_offset_ = 0;
+  byte *buffer_ = nullptr;
+
+  // cache of the header so that it can be automatically filled into new pages.
+  uint32_t table_oid_;
+  uint32_t version_;
+  uint32_t flags_;
+
+  /**
+   * Wirte a piece of data into the buffer. If the buffer is not enough, persist current buffer and allocate a new one.
+   * Note that alignment is not dealt in this function. It has to be done manually.
+   * @param data to be written.
+   * @param size of the data.
+   */
+  void WriteDataToBuffer(const byte *data, uint32_t size) {
+    uint32_t left_to_write = size;
+    while (left_to_write > 0) {
+      uint32_t remaining_buffer = block_size_ - page_offset_;
+      if (left_to_write >= remaining_buffer) {
+        memcpy(buffer_ + page_offset_, data, remaining_buffer);
+        page_offset_ += remaining_buffer;
+        left_to_write -= remaining_buffer;
+        data = data + remaining_buffer;
+        Persist();
+      } else {
+        memcpy(buffer_ + page_offset_, data, left_to_write);
+        page_offset_ += left_to_write;
+        return;
+      }
+    }
+  }
+
+  void ResetBuffer() {
+    CheckpointFilePage::Initialize(reinterpret_cast<CheckpointFilePage *>(buffer_));
+    page_offset_ = sizeof(CheckpointFilePage);
+    AlignBufferOffset<uint64_t>();  // align for ProjectedRow
+    GetPage()->SetVersion(version_);
+    GetPage()->SetTableOid(catalog::table_oid_t(table_oid_));
+    GetPage()->SetFlags(flags_);
+  }
+
+  void PersistBuffer() {
+    // TODO(zhaozhe): calculate CHECKSUM. Currently using default 0 as checksum
+    if (page_offset_ == sizeof(CheckpointFilePage)) {
+      // If the buffer has no contents, just return
+      return;
+    }
+    AlignBufferOffset<uint64_t>();
+    if (block_size_ - page_offset_ > sizeof(uint32_t)) {
```

Comment:
What happens in the case where (block_size_ - page_offset_) <= sizeof(uint32_t). In BufferedTupleWriter::SerializeTuple you guys only check (page_offset_ + checkpoint_record_size > block_size_) which doesn't take into account these extra 32 bits that need to be zeroed. So it's possible that for some block that this zeroing out won't be happening at the end of some blocks.